src/python/pants/base/specs.py

```diff
@@ -70,40 +79,42 @@ def all_address_target_pairs(cls, address_families):
     return addr_tgt_pairs
 
   @abstractmethod
-  def make_glob_patterns(self, address_mapper):
+  def make_glob_patterns(self, address_mapper: "AddressMapper") -> List[str]:
     """Generate glob patterns matching exactly all the BUILD files this spec covers."""
 
   @classmethod
-  def globs_in_single_dir(cls, spec_dir_path, address_mapper):
+  def globs_in_single_dir(cls, spec_dir_path: str, address_mapper: "AddressMapper") -> List[str]:
     """Implementation of `make_glob_patterns()` which only allows a single base directory."""
     return [os.path.join(spec_dir_path, pat) for pat in address_mapper.build_patterns]
 
 
 @dataclass(frozen=True)
 class SingleAddress(Spec):
   """A Spec for a single address."""
-  directory: Any
-  name: Any
+  directory: str
+  name: str
 
-  def __post_init__(self):
-    if self.directory is None or self.name is None:
-      raise ValueError(
-        f'A SingleAddress must have both a directory and name. Got: {self.directory}:{self.name}'
-      )
+  def __post_init__(self) -> None:
+    if self.directory is None:
+      raise ValueError(f'A SingleAddress must have a directory. Got: {self}')
+    if self.name is None:
+      raise ValueError(f'A SingleAddress must have a name. Got: {self}')
```

Comment:
This was split into two different checks because MyPy infers `self.directory is None` to always be `False` given the type hint that `directory` is `str` and not `Optional[str]`.

Also, I changed to print `self` because the dataclass will give really nice output already.